enable stack minimization for all crash dumps

bmc-msft opened the pull request:
This builds upon #591 to expand the stack minimization to crash reporting mechanisms.

Example (see #703 for an example without the new functionality):

```
$ onefuzz-agent local test-input /tmp/fuzz.exe /etc/passwd
{
  "crash_report": {
    "input_sha256": "a35b3ce1038750e9175a6dcd3f64c8d4e85720affb12cc11f5d0b6889274d06e",
    "executable": "/tmp/fuzz.exe",
    "crash_type": "SIGABRT",
    "crash_site": "0x7ffff7e0d18b in gsignal+0xcb (/usr/lib/x86_64-linux-gnu/libc-2.31.so+0x4618b)",
    "call_stack": [
      "#0 0x7ffff7e0d18b in gsignal+0xcb (/usr/lib/x86_64-linux-gnu/libc-2.31.so+0x4618b)",
      "#1 0x7ffff7dec859 in abort+0x12b (/usr/lib/x86_64-linux-gnu/libc-2.31.so+0x25859)",
      "#2 0x7ffff7e573ee in <unknown> (/usr/lib/x86_64-linux-gnu/libc-2.31.so+0x903ee)",
      "#3 0x7ffff7ef9b4a in __fortify_fail+0x2a (/usr/lib/x86_64-linux-gnu/libc-2.31.so+0x132b4a)",
      "#4 0x7ffff7ef83e6 in __chk_fail+0x16 (/usr/lib/x86_64-linux-gnu/libc-2.31.so+0x1313e6)",
      "#5 0x7ffff7ef7e09 in __strncpy_chk+0x19 (/usr/lib/x86_64-linux-gnu/libc-2.31.so+0x130e09)",
      "#6 0x400a54 in from_file+0xa4 (/tmp/fuzz.exe+0xa54)",
      "#7 0x7ffff7dee0b3 in __libc_start_main+0xf3 (/usr/lib/x86_64-linux-gnu/libc-2.31.so+0x270b3)",
      "#8 0x40077a in _start+0x2a (/tmp/fuzz.exe+0x77a)"
    ],
    "call_stack_sha256": "99625a7c103136e02910b65c7b60f1bbd1a7612242d6838da52d968369039409",
    "minimized_stack": [
      "__fortify_fail",
      "__chk_fail",
      "from_file"
    ],
    "minimized_stack_sha256": "237f13bfa384c6c2bc06369099373efbb36995a9ad00fd5469d354b5fc672ba1",
    "minimized_stack_function_names": [
      "__fortify_fail",
      "__chk_fail",
      "from_file"
    ],
    "minimized_stack_function_names_sha256": "237f13bfa384c6c2bc06369099373efbb36995a9ad00fd5469d354b5fc672ba1",
    "asan_log": "",
    "task_id": "00000000-0000-0000-0000-000000000000",
    "job_id": "00000000-0000-0000-0000-000000000000"
  }
}
$
```

bmc-msft commented:
Note, there is work to do on the windows side, primarily exposing the information in DebugStackFrame.

However, this is a first pass:

```
> 
onefuzz-agent.exe local test-input C:\Users\USERNAME\projects\test-artifacts\windows-trivial-crash\fuzz.exe .\Cargo.lock
{
  "crash_report": {
    "input_sha256": "24c3bb3be2e50c36320b9e737f81438e0ec34835ce39def8bec9921b64ac17e8",
    "executable": "C:\\Users\\USERNAME\\projects\\test-artifacts\\windows-trivial-crash\\fuzz.exe",
    "crash_type": "WriteAccessViolation",
    "crash_site": "fuzz!strncpy minkernel\\crts\\ucrt\\src\\appcrt\\string\\amd64\\strncpy.asm:46",
    "call_stack": [
      "fuzz!strncpy minkernel\\crts\\ucrt\\src\\appcrt\\string\\amd64\\strncpy.asm:46",
      "fuzz!check D:\\a\\onefuzz\\onefuzz\\src\\integration-tests\\trivial-crash\\fuzz.c:21",
      "fuzz!from_file D:\\a\\onefuzz\\onefuzz\\src\\integration-tests\\trivial-crash\\fuzz.c:67",
      "<corrupt frame(s)>"
    ],
    "call_stack_sha256": "8c0026819742a73973b587fd9e18545b1fd2770ebd65b9141f24bcc6c4e47b9c",
    "minimized_stack": [
      "fuzz!strncpy",
      "fuzz!check",
      "fuzz!from_file"
    ],
    "minimized_stack_sha256": "58cd174f108dcd9cf568209a818dcae72d4fb7b0b85037598e26c214f3c44fce",
    "minimized_stack_function_names": [
      "fuzz!strncpy",
      "fuzz!check",
      "fuzz!from_file"
    ],
    "minimized_stack_function_names_sha256": "58cd174f108dcd9cf568209a818dcae72d4fb7b0b85037598e26c214f3c44fce",
    "asan_log": "",
    "task_id": "00000000-0000-0000-0000-000000000000",
    "job_id": "00000000-0000-0000-0000-000000000000"
  }
}
```

bmc-msft commented:
Note, this is what the report looked like with just #703 on windows:

```
{
  "crash_report": {
    "input_sha256": "1e60a47ad5838f06f41d91d8dbb530316877f297aca7006ac516ff70c18c34cc",
    "executable": "C:\\Users\\bcaswell\\projects\\test-artifacts\\windows-trivial-crash\\fuzz.exe",
    "crash_type": "WriteAccessViolation",
    "crash_site": "fuzz!strncpy minkernel\\crts\\ucrt\\src\\appcrt\\string\\amd64\\strncpy.asm:46",
    "call_stack": [
      "fuzz!strncpy minkernel\\crts\\ucrt\\src\\appcrt\\string\\amd64\\strncpy.asm:46",
      "fuzz!check D:\\a\\onefuzz\\onefuzz\\src\\integration-tests\\trivial-crash\\fuzz.c:21",
      "fuzz!from_file D:\\a\\onefuzz\\onefuzz\\src\\integration-tests\\trivial-crash\\fuzz.c:67",
      "<corrupt frame(s)>"
    ],
    "call_stack_sha256": "8c0026819742a73973b587fd9e18545b1fd2770ebd65b9141f24bcc6c4e47b9c",
    "minimized_stack": [],
    "minimized_stack_sha256": null,
    "minimized_stack_function_names": [],
    "minimized_stack_function_names_sha256": null,
    "task_id": "00000000-0000-0000-0000-000000000000",
    "job_id": "00000000-0000-0000-0000-000000000000"
  }
}
```

bmc-msft commented:
As mentioned above, the windows function name is module!symbol.

ranweiler commented:
In your output report examples, the value of the `minimized_stack_function_names` field is always the same as the `minimized_stack` field. When would we expect them to differ?

ranweiler commented on src/agent/onefuzz/src/input_tester.rs:
```diff
@@ -217,9 +233,21 @@ impl<'a> Tester<'a> {
+                    })
                     .collect();
 
+                println!("GOT {:?}", call_stack);
```
```suggestion
```
May want to do a quick grep.

ranweiler commented on src/agent/onefuzz/src/input_tester.rs:
```diff
@@ -228,11 +256,14 @@ impl<'a> Tester<'a> {
-                    crash_type,
-                    crash_site,
-                })
+                let text = "".to_string();
```
Hmm, in `CrashLog`, should `text` maybe be wrapped in an `Option<_>`?

bmc-msft replied:
yes, eventually.  it maps to asan_log via a long chain of events, which isn't optional at the moment.  My thought was that should be a follow-on PR.

ranweiler replied:
SGTM!

ranweiler commented on src/agent/onefuzz/src/input_tester.rs:
```diff
@@ -305,36 +335,34 @@ impl<'a> Tester<'a> {
-                None
-            };
+            if crash_log.is_none() {
+                crash_log = if let Some(asan_dir) = &asan_dir {
```
Due to all the conditional mutating, how about a comment to the effect of:
```suggestion
                // If we didn't already capture a crash log directly, try to fall back on searching
                // for saved log files in a user-provided ASAN log directory (if present).
                crash_log = if let Some(asan_dir) = &asan_dir {
```

ranweiler replied:
Ah, that comment as written is wrong (we check there first, and fall back on stderr). But you get the idea.

bmc-msft commented:
New output, addressing comments : 
```json
{
  "crash_report": {
    "input_sha256": "8abb9b1b07f17b5442c5d71b38eaf599ddd37fa38a3062c709a7e68412e57922",
    "executable": "/home/USERNAME/projects/onefuzz/test-artifacts/linux-trivial-crash/fuzz.exe",
    "crash_type": "SIGABRT",
    "crash_site": "0x7ffff7e0d18b in gsignal+0xcb (/usr/lib/x86_64-linux-gnu/libc-2.31.so+0x4618b)",
    "call_stack": [
      "#0 0x7ffff7e0d18b in gsignal+0xcb (/usr/lib/x86_64-linux-gnu/libc-2.31.so+0x4618b)",
      "#1 0x7ffff7dec859 in abort+0x12b (/usr/lib/x86_64-linux-gnu/libc-2.31.so+0x25859)",
      "#2 0x7ffff7e573ee in <unknown> (/usr/lib/x86_64-linux-gnu/libc-2.31.so+0x903ee)",
      "#3 0x7ffff7ef9b4a in __fortify_fail+0x2a (/usr/lib/x86_64-linux-gnu/libc-2.31.so+0x132b4a)",
      "#4 0x7ffff7ef83e6 in __chk_fail+0x16 (/usr/lib/x86_64-linux-gnu/libc-2.31.so+0x1313e6)",
      "#5 0x7ffff7ef7e09 in __strncpy_chk+0x19 (/usr/lib/x86_64-linux-gnu/libc-2.31.so+0x130e09)",
      "#6 0x400a54 in from_file+0xa4 (/home/USERNAME/projects/onefuzz/test-artifacts/linux-trivial-crash/fuzz.exe+0xa54)",
      "#7 0x7ffff7dee0b3 in __libc_start_main+0xf3 (/usr/lib/x86_64-linux-gnu/libc-2.31.so+0x270b3)",
      "#8 0x40077a in _start+0x2a (/home/USERNAME/projects/onefuzz/test-artifacts/linux-trivial-crash/fuzz.exe+0x77a)"
    ],
    "call_stack_sha256": "768b179b2f645f838c62d13f174d28852cb6637e9f59ebfcfe23a1370537787c",
    "minimized_stack": [
      "#2 0x7ffff7e573ee in <unknown> (/usr/lib/x86_64-linux-gnu/libc-2.31.so+0x903ee)",
      "#6 0x400a54 in from_file+0xa4 (/home/USERNAME/projects/onefuzz/test-artifacts/linux-trivial-crash/fuzz.exe+0xa54)"
    ],
    "minimized_stack_sha256": "57ddb2a20635188293e08882d9f9d24a619cdfd73711ababa4b17e3212a1a6c5",
    "minimized_stack_function_names": [
      "from_file"
    ],
    "minimized_stack_function_names_sha256": "710dc924cf18883634a64c005c2b88f0199c041c10fa3fa360f62ca3a83de2da",
    "asan_log": "",
    "task_id": "00000000-0000-0000-0000-000000000000",
    "job_id": "00000000-0000-0000-0000-000000000000"
  }
}
```
